designs/2024-per-rule-autofix-configuration/README.md

```diff
@@ -0,0 +1,406 @@
+
+## Summary
+
+<!-- One-paragraph explanation of the feature. -->
+
+This RFC proposes adding support for controlling autofix behavior on a per-rule basis through shareable configuration.
+
+## Motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+
+Some rules support autofixing, which is often convenient, but in certain cases 
+the fixes may be broken, unsafe, or simply undesirable. \
+Ideally, unsafe autofixes should be treated as suggestions, and broken fixes should be reported. 
+
+However, ESLint is a large ecosystem, and some useful plugins are no longer actively maintained. 
+Even in actively maintained projects, users may want to disable autofixing for specific rules
+due to project-specific or personal preferences.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+### Abstract
+
+We introduce an *extended rule configuration format*:
+
+```ts
+type RuleEntry = Partial<{
+  /**
+   * The rule severity.
+   * 
+   * @default 2
+   */
+  severity: 0 | 1 | 2 | 'off' | 'warn' | 'error';
+
+  /**
+   * Array of options passed to the rule.
+   * 
+   * @default []
+   */
+  options: any[]
+
+  /**
+   * Whether to allow a rule to perform autofixes.
+   * 
+   * @default true
+   */
+  autofix: boolean;
+}>;
+```
+
+This format enables explicit control over a rule's autofix behavior,
+and opens the door to further extension with other meta-properties
+([example](https://github.com/eslint/eslint/issues/19342)).
+
+> [!IMPORTANT]
+> This is an additional configuration format; existing configurations remain valid and do not need to be rewritten.
+
+#### Autofix details
+
+* The `autofix` option defaults to `true`, maintaining current behavior 
+  (if a rule provides a fixer, it will run when `--fix` is used);
+* If autofix is disabled for a rule, its fix becomes a suggestion labeled 
+  "Apply disabled autofix" and can be applied manually through editor hints;
+* If the rule does not define a fixer, the `autofix` key has no effect.
+
+#### Additional extensions
+
+We extend `linterOptions.reportUnusedDisableDirectives` 
+and `linterOptions.reportUnusedInlineConfigs`(?) to support new extended format:
+
+```js
+export default [
+  {
+    linterOptions: {
+      reportUnusedDisableDirectives: {
+        severity: 'error',
+        autofix: false,
+      },
+    },
+  },
+];
+```
+
+We extend CLI engine to support new format as an additional option to existing ones:
+
+```bash
+npx eslint --fix --rule 'no-var: { autofix: false }' 
+```
+
+We extend directive parser engine to support new format as well:
+
+```js
+/* eslint eqeqeq: "off", curly: { severity: "warn", autofix: false } */
```

Comment:
I'm not 100% sure this is technically possible or desirable. Autofixing is CLI-specific and files can be linted without the CLI (i.e., in the browser), in which case this doesn't mean anything.

Need a few more opinions on this.

Reply:
I think it's okay to support this, to allow rules to be configured the same way everywhere.